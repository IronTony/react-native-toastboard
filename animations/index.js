// @flow
export * from "./BaseAnimation";

export * from "./Opacity";

export * from "./SlideRight";
export * from "./SlideLeft";
export * from "./SlideUp";
