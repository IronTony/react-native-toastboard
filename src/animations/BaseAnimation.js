// @flow
import { Animated } from "react-native";

interface AnimationConfig {
	easing?: (value: number) => number;
	deceleration?: number;
	duration?: number;
	delay?: number;
}

export interface BaseAnimationInterface {
	forward: () => Promise<void>;
	backward: () => Promise<void>;
	getAnimation: () => any;
}

export class BaseAnimation {
	config: AnimationConfig = {
		duration: 250,
		useNativeDriver: true
	};

	from: number;
	to: number;

	value: any;

	constructor(from: number, to: number, config?: AnimationConfig) {
		config !== undefined && (this.config = config);
		this.from = from;
		this.to = to;

		this.value = new Animated.Value(this.from);
	}
}
