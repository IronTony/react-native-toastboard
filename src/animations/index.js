// @flow
export * from "./BaseAnimation";

export * from "./Opacity";
export * from "./SlideUp";
