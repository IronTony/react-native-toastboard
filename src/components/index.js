// @flow
export * from "./Toast";
export * from "./Toaster";
