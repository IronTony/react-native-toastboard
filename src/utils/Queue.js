// @flow
export class Queue {
	active = false;
	list: Array<any> = [];

	onIteration: (item: any) => Promise<any> | void;
	beforeIteration: () => Promise<any> | void;
	afterIteration: () => Promise<any> | void;

	constructor(props: {
		onIteration?: () => Promise<any> | void;
		beforeIteration?: () => Promise<any> | void;
		afterIteration?: () => Promise<any> | void;
	} = {}) {
		if (typeof props.onIteration === "function") {
			this.onIteration = props.onIteration;
		}

		if (typeof props.beforeIteration === "function") {
			this.beforeIteration = props.beforeIteration;
		}

		if (typeof props.afterIteration === "function") {
			this.afterIteration = props.afterIteration;
		}

	}

	push = (item: any) => {
		this.list.push(item);
	}

	process = async () => {
		this.active = true;

		if (this.beforeIteration) {
			await this.beforeIteration();
		}

		if (!this.list.length) {
			this.active = false;
			return;
		}

		if (this.onIteration) {
			await this.onIteration(this.list[0]);
		}

		this.list = this.list.slice(1);

		if (this.afterIteration) {
			await this.afterIteration();
		}

		this.process();
	}

	start = () => {
		if (this.active) {
			return;
		}

		this.process();
	}
}
