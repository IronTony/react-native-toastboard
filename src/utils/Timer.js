// @flow
export class Timer {
	stop: () => void = () => { };
	pause: () => void = () => { };
	resume: () => void = () => { };

	/* eslint-disable-next-line no-magic-numbers */
	start = (delay: number = 2000): Promise<void> => {
		return new Promise((resolve) => {
			if (delay <= 0) {
				return resolve();
			}

			const time = Date.now();
			let remaining = delay;

			let timeout = setTimeout(resolve, delay);

			this.stop = () => {
				resolve();
				clearTimeout(timeout);

				this.pause = () => { };
				this.resume = () => { };
			};

			this.pause = () => {
				remaining -= Date.now() - time;
				if (remaining <= 0) {
					return this.stop();
				}

				clearTimeout(timeout);
			};

			this.resume = () => {
				if (remaining <= 0) {
					return this.stop();
				}

				timeout = setTimeout(this.stop, remaining);
			};
		});
	}
}
