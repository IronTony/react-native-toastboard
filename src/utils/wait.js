// @flow
/* eslint-disable-next-line no-magic-numbers */
export function wait( msec: number = 2000): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, msec));
}
