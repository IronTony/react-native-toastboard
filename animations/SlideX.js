// @flow
import { Animated } from "react-native";
import { BaseAnimation } from "./BaseAnimation";
import type { BaseAnimationInterface } from "./BaseAnimation";

export class SlideX extends BaseAnimation implements BaseAnimationInterface {
	forward = (): Promise<void> => {
		return new Promise((resolve) => {
			Animated.timing(this.value, {
				toValue: this.to,
				...this.config
			}).start(resolve);
		});
	}

	backward = (): Promise<void> => {
		return new Promise((resolve) => {
			Animated.timing(this.value, {
				toValue: this.from,
				...this.config
			}).start(resolve);
		});
	}

	getAnimation = () => {
		return {
			transform: [{ translateX: this.value }]
		};
	}
}
