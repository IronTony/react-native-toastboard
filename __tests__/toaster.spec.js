/* eslint-disable no-undef */
import React from "react";
import { Text } from "react-native";
import renderer from "react-test-renderer";

import { Toaster } from "../src";
import { BaseAnimation } from "../animations";

import { wait } from "../src/utils/wait";

const consoleOrigin = global.console;
const mockAnimation = (new class extends BaseAnimation {
    getAnimation = () => 1;
    forward = () => { };
    backward = () => { };
});

describe("<Toaster />", () => {
    beforeAll(() => {
        global.console = {
            ...consoleOrigin,
            warn: () => { }
        };
    });

    afterAll(() => {
        global.console = consoleOrigin;
    });

    test("Should render without <TouchableOpacity>", () => {
        const tree = renderer.create(<Toaster />);

        expect(tree.toJSON()).toMatchSnapshot();

        tree.unmount();
    });

    test("Should render with <TouchableOpacity>", () => {
        const tree = renderer.create(<Toaster hideOnPress />);

        expect(tree.toJSON()).toMatchSnapshot();

        tree.unmount();
    });

    test("Should render specified children instead default <Toast />", async () => {
        const tree = renderer.create(
            <Toaster animation={mockAnimation}>
                {({ message }) => <Text>{message}</Text>}
            </Toaster>
        );

        Toaster.info("test", 0);

        expect(tree.toJSON()).toMatchSnapshot();

        tree.unmount();
    });

    test("Should apply specified middleware", () => {
        const tree = renderer.create(
            <Toaster
                animation={mockAnimation}
                middleware={() => "transformed"}
            />
        );

        Toaster.info("test", 0);

        expect(tree.toJSON()).toMatchSnapshot();

        tree.unmount();
    });

    test("Should warn when try to create message if container not rendered", () => {
        const spy = jest.spyOn(console, "warn");

        Toaster.info("");
        Toaster.debug("");
        Toaster.error("");
        Toaster.success("");

        expect(spy).toHaveBeenCalledTimes(4);

        spy.mockClear();
    });

    test("Should re-define 'createToast' after mount", () => {
        const tree = renderer.create(
            <Toaster
                animation={mockAnimation}
            />
        );

        const spy = jest.spyOn(console, "warn");

        Toaster.info("");
        Toaster.debug("");
        Toaster.error("");
        Toaster.success("");

        expect(spy).not.toBeCalled();

        tree.unmount();
        spy.mockClear();
    });

    // test("Should execute 'nextItem' method if it exist onPress", () => {
    //     const instance = new Toaster({ animation: mockAnimation });
    //     instance.nextItem = () => { };

    //     const spy = jest.spyOn(instance, "nextItem");

    //     instance.handlePress();
    //     expect(spy).toHaveBeenCalledTimes(1);

    //     instance.handlePress();
    //     expect(spy).toHaveBeenCalledTimes(1);
    // });

    test("Should execute specified onShow/onHide props", async () => {
        let showed = false;
        let hided = false;

        const instance = new Toaster({
            animation: mockAnimation,
            onShow: () => { showed = true },
            onHide: () => { hided = true }
        });

        await instance.handleIteration({ duration: 0 });

        expect(showed).toBeTruthy();
        expect(hided).toBeTruthy();
    });

    test("Should add additional timeout specified in delayBetween prop", async () => {
        const spy = jest.spyOn(require("../src/utils/wait"), "wait");

        const instance = new Toaster({
            animation: mockAnimation,
            delayBetween: 1
        });

        await instance.handleIteration({ duration: 0 });

        expect(spy).toHaveBeenCalledTimes(1);
        spy.mockClear();
    });

    // test("Should define 'nextItem' method on each iteration", async () => {
    //     const instance = new Toaster({
    //         animation: mockAnimation
    //     });

    //     instance.handleIteration({ duration: 2 });
    //     await wait(1);
    //     expect(instance.nextItem).toBeInstanceOf(Function);

    //     instance.nextItem();
    //     await wait(1);

    //     expect(instance.nextItem).toBeUndefined();
    // });
});
