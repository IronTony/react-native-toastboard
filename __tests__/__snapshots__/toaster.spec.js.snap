// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toaster /> Should render specified children instead default <Toast /> 1`] = `
<View
  style={
    Object {
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 1,
    }
  }
>
  <Text>
    test
  </Text>
</View>
`;

exports[`<Toaster /> Should render with <TouchableOpacity> 1`] = `
<View
  style={
    Object {
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateY": -57,
        },
      ],
      "zIndex": 1,
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  />
</View>
`;

exports[`<Toaster /> Should render without <TouchableOpacity> 1`] = `
<View
  style={
    Object {
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateY": -57,
        },
      ],
      "zIndex": 1,
    }
  }
/>
`;
